File: Behavioral/State/Java/TrafficLight.java
// TrafficLight.java

public class TrafficLight {
  private TrafficLightState state;

  public TrafficLight() {
    state = new RedLight(); // Initial state
  }

  public void setState(TrafficLightState state) {
    this.state = state;
  }

  public void change() {
    state.handle(this);
  }
}

File: Behavioral/State/Cpp/TrafficLight.cpp
// TrafficLight.cpp

#include "TrafficLight.h"
#include "RedLight.h"

TrafficLight::TrafficLight() {
  state = new RedLight(); // Initial state
}

TrafficLight::~TrafficLight() {
  delete state;
}

void TrafficLight::setState(TrafficLightState* newState) {
  delete state;
  state = newState;
}

void TrafficLight::change() {
  state->handle(this);
}
